File: codegen/codegen-test/src/test/resources/snapshots/erlang/server/weather_server.erl
-module(weather_server).
-export([handle_get_current_time/2, handle_get_forecast/2]).

-spec handle_get_current_time(Req :: term(), State :: term()) -> {ok, term()} | {error, term()}.
handle_get_current_time(Req, State) ->
    {error, not_implemented}.

-callback get_current_time(Input :: get_current_time_input(), Context :: term()) ->
    {ok, get_current_time_output()} | {error, term()}.


-spec handle_get_forecast(Req :: term(), State :: term()) -> {ok, term()} | {error, term()}.
handle_get_forecast(Req, State) ->
    {error, not_implemented}.

-callback get_forecast(Input :: get_forecast_input(), Context :: term()) ->
    {ok, get_forecast_output()} | {error, no_such_resource_error()}.


-behaviour(smithy_handler).

File: codegen/codegen-test/src/test/resources/snapshots/weather/elixir/server/weather_server.ex
defmodule Weather.Server do
  @behaviour SmithyHandler
  @spec handle_get_current_time(request :: term(), state :: term()) :: {:ok, term()} | {:error, term()}
  def handle_get_current_time(_input, _context) do
    {:error, :not_implemented}
  end

  @callback get_current_time(input :: GetCurrentTimeInput.t(), context :: term()) ::
    {:ok, GetCurrentTimeOutput.t()} | {:error, term()}

  @spec handle_get_forecast(request :: term(), state :: term()) :: {:ok, term()} | {:error, term()}
  def handle_get_forecast(_input, _context) do
    {:error, :not_implemented}
  end

  @callback get_forecast(input :: GetForecastInput.t(), context :: term()) ::
    {:ok, GetForecastOutput.t()} | {:error, NoSuchResourceError.t()}

  @impl SmithyHandler
  def handle_request(operation, input, context) do
    case operation do
      :get_current_time -> handle_get_current_time(input, context)
      :get_forecast -> handle_get_forecast(input, context)
      _ -> {:error, :not_found}
    end
  end
end
